counsel.el

```diff
@@ -4918,6 +4918,36 @@ An extra action allows to switch to the process buffer."
 ;; `counsel-slime-repl-history' within
 ;; `counsel--browse-history-action'.
 
+;;** `counsel-esh-dir-history'
+(declare-function eshell/cd "eshell")
+
+(defun counsel--esh-dir-history-action-cd (pair)
+  "Default action for counsel-esh-dir-history. It changes the
+current working directory to the one selected by the user."
+  (eshell/cd (car pair)))
+
+(defun counsel--esh-dir-history-action-edit (pair)
+  "Action for counsel-esh-dir-history to insert the selected
+directory to the Eshell buffer prefixed by \"cd \", allowing the
+caller to modify parts of the directory before switching to it."
+  (ivy--action-insert (format "cd %s" (car pair))))
+
+(defvar eshell-last-dir-ring)
+
+;;;###autoload
+(cl-defun counsel-esh-dir-history ()
+  "Use Ivy to navigate and jump through Eshell's directory stack."
+  (interactive)
+  (require 'em-dirs)
+  (ivy-read "Directory to change to: " (ivy-history-contents eshell-last-dir-ring)
```

Comment:
Would it be possible to reuse `counsel--browse-history`?

Reply:
Do you mean calling `counsel--browse-history` instead of doing my own call to `ivy-read`? Probably I'm missing something but wouldn't that prevent me from being able to... ?

1) Customize the prompt string (I find it useful that the prompt gives a hint of what the default action does, but perhaps this is not a common practice that I should avoid)
2) Set the default action inside the declaration of `counsel-esh-dir-history`.

Perhaps 2) could be addressed by setting the default action later on somehow.

Reply:
There's nothing stopping you from extending `counsel--browse-history` to suit your needs ;).

That said, I'm not familiar with `counsel--browse-history`'s design. Hence my question.

Reply:
I've patched `counsel--browse-history` to allow customising the prompt and the default action. Please take a look.